Write RentsController #create & #index

cquiro opened the pull request:
## Summary
- Implement counter cache for user's rents
- Add returned_at column to rents table
- Create user and rent serializers
- Create RentsController with create and index actions
- Write routes for RentsController #create and #index 
- Write tests for the new RentsController
- Update tests for Rent model


## Trello Card
https://trello.com/c/aGMkNK4Y/31-crear-endpoint-de-creacion-de-rent-e-index

## Notes

This PR includes two migrations.

cquiro committed:
Create rents counter cache for user

- Write and run migration to add column rents_count to user
and update counters for existing users

- Add counter_cache: true to the 'belongs_to :user'
relationship in the Rent model

cquiro committed:
Add returned_at column to rents table.

cquiro committed:
Create endpoint for RentsController#Index

- Create RentsController with index action

- Set up route for api/v1/users/:user_id/rents

- Write/update serializers for rent and user

cquiro committed:
Return start_date and end_date instead of from and to in RentSerializer.

cquiro committed:
Add endpoint to create a rent for a user.

cquiro committed:
Update rent_spec.rb to check the user's rents_count is updated when a new rent is created.

cquiro committed:
Add specs for index and create actions in RentsController.

cquiro committed:
Fix rubocop offense in rent_spec.rb

lmasello commented on app/controllers/api/v1/rents_controller.rb:
```diff
@@ -0,0 +1,39 @@
+      private
+
+      def new_rent
+        @new_rent ||= user.rents.new(create_rent_params)
```
No need to use memorizaiton here. I'd write `user.rents.new(create_rent_params)` above in the create method.

facucompe replied:
I would do this in the create method

lmasello commented on db/schema.rb:
```diff
@@ -51,6 +52,7 @@
     t.string "uid", default: "", null: false
     t.boolean "allow_password_change", default: false
     t.json "tokens"
+    t.integer "rents_count", default: 0, null: false
```
We should add a model validation, so that rents_count it's always >= 0.

lmasello commented on spec/controllers/api/v1/rents_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+        rents, each_serializer: RentSerializer, root: false
+      ).to_json
+
+      expect(response_body.to_json) =~ JSON.parse(expected)
```
Doesn't it work with `.to eq`?

facucompe replied:
Try using `expect(response_body).to have_been_serialized_with(RentSerializer)`

cquiro replied:
@lmasello @facucompe I couldn't make it work with any of the two approaches.  It could be that I'm missing some extra code so that `have_been_serialized_with` works (a gem or a helper maybe?).

The closest thing I found was this https://robots.thoughtbot.com/validating-json-schemas-with-an-rspec-matcher. Should I do it that way?

lmasello replied:
It's also a nice approach. However, instead of adding another gem let's take a look at the contents of `response_body`. 

If we insert a debugger (e.g., `byebug`) before this expect, we see that `response_body` contains the expected result but as a value of the key `page` (we may guess that this is because of the pagination, right?).

So, something like: 
`expect(response_body['page'].to_json).to eq expected` should work as well as facu's suggested approach. You can use whichever you like :)

cquiro replied:
Thanks! My bad. I didn't check the response properly. I will use the `.to  eq` approach. I did like the `have_been_serialized_with` approach, but I asked around and turns out it's a Wolox custom matcher.

lmasello commented on spec/controllers/api/v1/rents_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  describe 'POST #create' do
+    let!(:book) { create(:book) }
```
Use lazy load (e.g., `let(:book) ...`).
You can read ["Use let and let!"](http://www.betterspecs.org).

lmasello commented on spec/controllers/api/v1/rents_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+
+    it 'responds with the serialized new rent' do
+      expect(response_body['start_date'].to_date).to eq rent_attrs[:start_date]
+      expect(response_body['book']['id']).to eq book.id
```
Careful with this. 
> Each test should make only one assertion.
[¨Single expectation test¨](http://www.betterspecs.org/)

cquiro replied:
@lmasello I did read that, but I couldn't think of a better way to write this test with a single expectation. Should I pick one of the three expectations and test only for that?

lmasello replied:
Or you can test them separately, like:
`it 'responds with the right book'`
And similar with the user.

cquiro replied:
Good. I'll do that.

lmasello commented on db/migrate/20180926133830_add_rents_count_to_users.rb:
```diff
@@ -0,0 +1,18 @@
+
+  def data
+    execute <<-SQL.squish
+      UPDATE users
```
It's preferable to use `find_each` instead of a SQL block.

cquiro committed:
Validate rents_count >= 0 in the User model

- Add validation to ensure rents_count is always >= 0

- Create user_spec.rb and test new validation

cquiro committed:
Replace let! with let.

cquiro committed:
Remove new_rent method in rents_controller.

cquiro committed:
Replace let with let! when creating rents in rents_controller_spec.rb

cquiro committed:
Use find_each instead of a SQL block in AddRentsCountToUser migration.

cquiro committed:
Merge branch 'master' into rent-create-index

cquiro committed:
Update the README.

cquiro committed:
Separate expectations in RentsControllerSpec create spec.

cquiro committed:
Replace '=~' with '.to eq' in the specs.

cquiro committed:
Rewrite reset rents_counter code in AddRentsCountToUsers migration

lmasello commented on db/migrate/20180926133830_add_rents_count_to_users.rb:
```diff
@@ -0,0 +1,19 @@
+    end
+  end
+
+  def data
```
Try to rename it to something more self descriptive

lmasello commented on db/migrate/20180926133830_add_rents_count_to_users.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  def data
+    ids = Set.new
```
You can just do: 
ids = Rent.all.map(&:user_id)

cquiro committed:
Rewrite users counter reset method in db/migrate/20180926133830_add_rents_count_to_users.rb

facucompe commented on db/migrate/20180926133830_add_rents_count_to_users.rb:
```diff
@@ -0,0 +1,17 @@
+  end
+
+  def reset_users_rents_counter
+    ids = Rent.all.map(&:user_id)
```
`Rent.find_each.map(&:user_id)`

cquiro committed:
Replace Rent.all.map(&:user_id) with Rent.find_each.map(&:user_id) in db/migrate/20180926133830_add_rents_count_to_users.rb
